### ISSUE TITLE
Multithreaded Tracking

### ISSUE THREAD

**Issue by codecodingg** [mikel-brostrom/yolo_tracking#1088]
### Search before asking

- [X] I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.

### Question

hello, I'm new to yolov8 tracking. in my case, I modified track.py load args from json file, it runs ok with one video, but two videos with threading only run few frame and stucked.  modified track.py files:

import argparse
from functools import partial
from pathlib import Path

import torch

from boxmot import TRACKERS
from boxmot.tracker_zoo import create_tracker
from boxmot.utils import ROOT, WEIGHTS
from boxmot.utils.checks import TestRequirements
from examples.detectors import get_yolo_inferer

__tr = TestRequirements()
__tr.check_packages(('ultralytics @ git+https://github.com/mikel-brostrom/ultralytics.git', ))  # install

from ultralytics import YOLO
from ultralytics.data.utils import VID_FORMATS
from ultralytics.utils.plotting import save_one_box

from examples.utils import write_mot_results

import cv2
import numpy as np
import json

import threading

def load_config():
    with open('config.json', 'r', encoding='utf-8') as f:
        config = json.load(f)
    with open('cameras.json', 'r', encoding='utf-8') as f:
        cameras = json.load(f)
    return config, cameras

def on_predict_start(predictor, persist=False):
    """
    Initialize trackers for object tracking during prediction.

    Args:
        predictor (object): The predictor object to initialize trackers for.
        persist (bool, optional): Whether to persist the trackers if they already exist. Defaults to False.
    """
    
    assert predictor.custom_args["tracking_method"] in TRACKERS, \
        f"'{predictor.custom_args['tracking_method']}' is not supported. Supported ones are {TRACKERS}"

    tracking_config = \
        ROOT /\
        'boxmot' /\
        'configs' /\
        (predictor.custom_args["tracking_method"] + '.yaml')
    
    trackers = []

    for i in range(predictor.dataset.bs):
        tracker = create_tracker(
            predictor.custom_args["tracking_method"],
            tracking_config,
            predictor.custom_args["reid_model"],
            predictor.device,
            predictor.custom_args["half"],
            predictor.custom_args["per_class"]
        )
        # motion only modeles do not have
        if hasattr(tracker, 'model'):
            tracker.model.warmup()
        trackers.append(tracker)

    predictor.trackers = trackers

@torch.no_grad()
def run(args, camera):
    yolo = YOLO(
        args["yolo_model"] if 'yolov8' in str(args["yolo_model"]) else 'yolov8n.pt',
    )
    results = yolo.track(
        source=args["source"],
        conf=args["conf"],
        iou=args["iou"],
        show=args["show"],
        stream=True,
        device=args["device"],
        show_conf=args["show_conf"],
        show_labels=args["show_labels"],
        save=args["save"],
        verbose=args["verbose"],
        exist_ok=args["exist_ok"],
        project=args["project"],
        name=args["name"],
        classes=args["classes"],
        imgsz=args["imgsz"]
    )

    yolo.add_callback('on_predict_start', partial(on_predict_start, persist=True))
    if 'yolov8' not in str(args["yolo_model"]):
        # replace yolov8 model
        print("replace yolov8 model")
        m = get_yolo_inferer(args["yolo_model"])
        model = m(
            model=args["yolo_model"],
            device=yolo.predictor.device,
            args=yolo.predictor.args
        )
        yolo.predictor.model = model

    # # store custom args in predictor
    yolo.predictor.custom_args = args
    
    for frame_idx, r in enumerate(results):
        print(camera["source"], frame_idx)
        if r.boxes.data.shape[1] == 7:

            if yolo.predictor.source_type.webcam or args["source"].endswith(VID_FORMATS):
                p = yolo.predictor.save_dir / 'mot' / (args["source"] + '.txt')
                yolo.predictor.mot_txt_path = p
            elif 'MOT16' or 'MOT17' or 'MOT20' in args["source"]:
                p = yolo.predictor.save_dir / 'mot' / (Path(args["source"]).parent.name + '.txt')
                yolo.predictor.mot_txt_path = p

            if args["save_mot"]:
                write_mot_results(
                    yolo.predictor.mot_txt_path,
                    r,
                    frame_idx,
                )

            if args["save_id_crops"]:
                for d in r.boxes:
                    print('args["save_id_crops', d.data)
                    save_one_box(
                        d.xyxy,
                        r.orig_img.copy(),
                        file=(
                            yolo.predictor.save_dir / 'crops' /
                            str(int(d.cls.cpu().numpy().item())) /
                            str(int(d.id.cpu().numpy().item())) / f'{frame_idx}.jpg'
                        ),
                        BGR=True
                    )
            # camera coordinates to 2D floor plan
            px, py, pw, ph = r.boxes.xywh.numpy()[0]
            print("px, py, pw, ph: ", px, py, pw, ph)
            point = np.array([[px, py * 2]], dtype='float32')
            point = np.array([point])
            pointOut = cv2.perspectiveTransform(point, camera["ho"])
            realx, realy = pointOut[0][0]
            # print("realx, realy: ", realx, realy)
            outPos = {
                "ip": camera["ip"],
                "x": realx.item(),
                "y": realy.item(),
                "track_id": r.boxes.id.int().cpu()[0].item(),
                "class": r.boxes.cls.int().numpy().tolist()
            }
            print("outPos: ", outPos)

    if args["save_mot"]:
        print(f'MOT results saved to {yolo.predictor.mot_txt_path}')

    
if __name__ == "__main__":
    config, cameras = load_config()
    config["yolo_model"] = Path(config["yolo_model"])
    config["reid_model"] = Path(config["reid_model"])
   
    threads = []

    for camera in cameras:
        if len(camera["perspective_mapping"]["pixle"]):
            pts_src = np.array(camera["perspective_mapping"]["pixle"])
            pts_dst = np.array(camera["perspective_mapping"]["real"])
            ho, status = cv2.findHomography(pts_src, pts_dst)
            camera["ho"] = ho
            config["source"] = camera["source"]
            print(f"run config: {config}")
            thread = threading.Thread(target=run, args=(config, camera), daemon=True)
            threads.append(thread)
    for th in threads:
        th.start()
    for th in threads:
        th.join()
    
## config.json flie content:
{
    "yolo_model": "/home/hdjw/yolo_tracking/examples/weights/yolov8n.pt",
    "reid_model": "/home/hdjw/yolo_tracking/examples/weights/osnet_x0_25_msmt17.pt",
    "tracking_method": "deepocsort",
    "imgsz": [640],
    "conf": 0.5,
    "iou": 0.7,
    "device": "0",
    "show":  false,
    "save":  false,
    "classes": [0],
    "project": "/home/hdjw/yolo_tracking/runs/track",
    "name": "exp", 
    "exist_ok": false, 
    "half": false, 
    "vid_stride": 1, 
    "show_labels": true, 
    "show_conf": true, 
    "save_txt": false, 
    "save_id_crops": false, 
    "save_mot": false, 
    "line_width": null, 
    "per_class": false, 
    "verbose": true
}

## cameras.json content
[
    {
        "ip": "192.168.99.106",
        "source": "./106v.mp4",
        "perspective_mapping": {
            "pixle": [[32, 489], [522, 191], [1341, 263], [923, 791]],
            "real": [[4.1, 2.3], [7.6, 2.3], [7.6, 6.7], [4.1, 6.7]]
        }
    },
    {
        "ip": "192.168.99.107",
        "source": "./107v.mp4",
        "perspective_mapping": {
            "pixle": [[618, 450], [591, 266], [1290, 231], [1556, 383]],
            "real": [[7.6, 2.3], [10.6, 2.3], [10.6, 6.7], [7.6, 6.7]]
        }
    }
]

## terminal output
run config: {'yolo_model': PosixPath('/home/hdjw/yolo_tracking/examples/weights/yolov8n.pt'), 'reid_model': PosixPath('/home/hdjw/yolo_tracking/examples/weights/osnet_x0_25_msmt17.pt'), 'tracking_method': 'deepocsort', 'imgsz': [640], 'conf': 0.5, 'iou': 0.7, 'device': '0', 'show': False, 'save': False, 'classes': [0], 'project': '/home/hdjw/yolo_tracking/runs/track', 'name': 'exp', 'exist_ok': False, 'half': False, 'vid_stride': 1, 'show_labels': True, 'show_conf': True, 'save_txt': False, 'save_id_crops': False, 'save_mot': False, 'line_width': None, 'per_class': False, 'verbose': True, 'source': './106v.mp4'}
run config: {'yolo_model': PosixPath('/home/hdjw/yolo_tracking/examples/weights/yolov8n.pt'), 'reid_model': PosixPath('/home/hdjw/yolo_tracking/examples/weights/osnet_x0_25_msmt17.pt'), 'tracking_method': 'deepocsort', 'imgsz': [640], 'conf': 0.5, 'iou': 0.7, 'device': '0', 'show': False, 'save': False, 'classes': [0], 'project': '/home/hdjw/yolo_tracking/runs/track', 'name': 'exp', 'exist_ok': False, 'half': False, 'vid_stride': 1, 'show_labels': True, 'show_conf': True, 'save_txt': False, 'save_id_crops': False, 'save_mot': False, 'line_width': None, 'per_class': False, 'verbose': True, 'source': './107v.mp4'}

2023-08-22 14:34:44.444 | SUCCESS  | boxmot.appearance.reid_model_factory:load_pretrained_weights:207 - Successfully loaded pretrained weights from "/home/hdjw/yolo_tracking/examples/weights/osnet_x0_25_msmt17.pt"
2023-08-22 14:34:44.444 | WARNING  | boxmot.appearance.reid_model_factory:load_pretrained_weights:211 - The following layers are discarded due to unmatched keys or layer size: ('classifier.weight', 'classifier.bias')
2023-08-22 14:34:44.454 | SUCCESS  | boxmot.appearance.reid_model_factory:load_pretrained_weights:207 - Successfully loaded pretrained weights from "/home/hdjw/yolo_tracking/examples/weights/osnet_x0_25_msmt17.pt"
2023-08-22 14:34:44.454 | WARNING  | boxmot.appearance.reid_model_factory:load_pretrained_weights:211 - The following layers are discarded due to unmatched keys or layer size: ('classifier.weight', 'classifier.bias')
./106v.mp4 0
video 1/1 (1/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 71.8ms
./107v.mp4 0
video 1/1 (1/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 72.5ms
./106v.mp4 1
video 1/1 (2/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 6.8ms
./107v.mp4 1
video 1/1 (2/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 7.6ms
./106v.mp4 2
video 1/1 (3/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 6.9ms
./107v.mp4 2
video 1/1 (3/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 7.5ms
./106v.mp4 3
video 1/1 (4/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 7.1ms
./107v.mp4 3
video 1/1 (4/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 7.3ms
./106v.mp4 4
video 1/1 (5/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 6.3ms
./107v.mp4 4
video 1/1 (5/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.8ms
./107v.mp4 5
video 1/1 (6/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.5ms
./106v.mp4 5
video 1/1 (6/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.7ms
./107v.mp4 6
video 1/1 (7/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.6ms
./106v.mp4 6
video 1/1 (7/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.6ms
./107v.mp4 7
video 1/1 (8/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.9ms
./106v.mp4 7
video 1/1 (8/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.9ms
./107v.mp4 8
video 1/1 (9/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.9ms
./106v.mp4 8
video 1/1 (9/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.9ms
./107v.mp4 9
video 1/1 (10/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 6.0ms
./106v.mp4 9
video 1/1 (10/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 6.1ms
./107v.mp4 10
video 1/1 (11/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.8ms
./106v.mp4 10
video 1/1 (11/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.7ms
./107v.mp4 11
video 1/1 (12/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.7ms
./106v.mp4 11
video 1/1 (12/13572) /home/hdjw/yolo_tracking/107v.mp4: 384x640 (no detections), 5.5ms

### could you give me some tips to find out what's wrong with it
